tool/planet/etcd.go

```diff
@@ -282,6 +287,91 @@ func etcdUpgrade(rollback bool) error {
 	return nil
 }
 
+// startWatchingEtcdMasters creates a control loop which polls etcd for the etcd cluster member list, and updates the
+// etcd gateway configuration with any changes. This keeps the etcd gateway load balancing in sync with the cluster.
+func startWatchingEtcdMasters(ctx context.Context) error {
+	config := etcdconf.Config{
+		Endpoints: []string{DefaultEtcdEndpoints},
+		KeyFile:   DefaultEtcdctlKeyFile,
+		CertFile:  DefaultEtcdctlCertFile,
+		CAFile:    DefaultEtcdctlCAFile,
+	}
+	cli, err := config.NewClientV3()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	go watchEtcdMasters(ctx, cli)
+	return nil
+}
+
+func watchEtcdMasters(ctx context.Context, client *etcdv3.Client) {
+	ticker := time.NewTicker(15 * time.Second)
+	defer ticker.Stop()
+
+	gateway := etcdGateway{}
+
+	for {
+		select {
+		case <-ticker.C:
+			err := gateway.resyncEtcdMasters(ctx, client)
+			if err != nil {
+				log.WithError(err).Warn("Error resyncing etcd master list.")
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+type etcdGateway struct {
+	clientURLs []string
+}
+
+func (e *etcdGateway) resyncEtcdMasters(ctx context.Context, client *etcdv3.Client) error {
+	memberList, err := client.MemberList(ctx)
+	if err != nil {
+		return trace.Wrap(err, "error retrieving member list")
+	}
+
+	newClientURLs := []string{}
+	for _, member := range memberList.Members {
+		memberURLs := member.GetClientURLs()
+		if len(memberURLs) == 0 {
+			return trace.Wrap(err, "etcd member doesn't have any client urls")
+		}
+
+		// Only use the first memberUrl to prevent the same member appearing multiple times
+		u, err := url.Parse(memberURLs[0])
+		if err != nil {
+			return trace.Wrap(err, "error parsing etcd member url").AddField("url", memberURLs[0])
+		}
+
+		newClientURLs = append(newClientURLs, u.Host)
+	}
+
+	// Only rewrite the configuration if there are changes
+	sort.Strings(newClientURLs)
+	if reflect.DeepEqual(newClientURLs, e.clientURLs) {
+		return nil
+	}
+
+	env := fmt.Sprintf("%v=%q", EnvEtcdGatewayEndpoints, strings.Join(newClientURLs, ","))
+	log.WithField("file", DefaultEtcdSyncedEnvFile).Info("Updating etcd gateway environment: ", env)
+	err = utils.SafeWriteFile(DefaultEtcdSyncedEnvFile, []byte(env), constants.SharedReadMask)
+	if err != nil {
+		return trace.Wrap(err, "failed to update etcd synced environment").AddField("file", DefaultEtcdSyncedEnvFile)
+	}
+
+	err = systemctl(ctx, "restart", ETCDServiceName)
```

Comment:
There's also a concept of a [path](https://www.freedesktop.org/software/systemd/man/systemd.path.html#Unit=) systemd unit which could be used to track changes to etcd-relevant files - i.e. it could watch other environment files and restart etcd on changes - it could be more consistent to do that way.